### ISSUE TITLE
Regression: Invalid type resolution of properties for types with generics

### ISSUE THREAD

**Issue by martin-sladecek** [jdbi/jdbi#1914]
This was caused by #1665.

When binding beans of a type with generics that has some property with generics, the property type is erased.
E.g.
```
public class MyEntity<T> {
    public List<String> getAbc();
}
```
Binding this type as a bean will resolve "abc" property as "List" type, losing the generic type. This is because `GenericTypeReflector.getExactReturnType` will erase the type if the bean type is "raw" (i.e. Class, not ParameterizedType), which is always the case in bound beans.
This might cause problems in custom ArgumentFactories that rely on this information.

**Comment by martin-sladecek** [comment]
The following workaround fixes the issue:
```
diff --git a/core/src/main/java/org/jdbi/v3/core/mapper/reflect/internal/BeanPropertiesFactory.java b/core/src/main/java/org/jdbi/v3/core/mapper/reflect/internal/BeanPropertiesFactory.java
index 3c08959c2..d14677e66 100644
--- a/core/src/main/java/org/jdbi/v3/core/mapper/reflect/internal/BeanPropertiesFactory.java
+++ b/core/src/main/java/org/jdbi/v3/core/mapper/reflect/internal/BeanPropertiesFactory.java
@@ -159,12 +159,21 @@ public class BeanPropertiesFactory {
 
                 return QualifiedType.of(
                     Optional.ofNullable(descriptor.getReadMethod())
-                        .map(m -> GenericTypeReflector.getExactReturnType(m, actualBeanType))
+                        .map(this::getExactReturnType)
                         .orElseGet(() -> GenericTypeReflector.getExactParameterTypes(descriptor.getWriteMethod(), actualBeanType)[0]))
                     .withAnnotations(
                         new Qualifiers().findFor(descriptor.getReadMethod(), descriptor.getWriteMethod(), setterParam));
             }
 
+            private Type getExactReturnType(Method m) {
+                Type exactReturnType = GenericTypeReflector.getExactReturnType(m, actualBeanType);
+                if (GenericTypeReflector.isMissingTypeParameters(exactReturnType)) {
+                    return GenericTypeReflector.getExactReturnType(m, GenericTypeReflector.addWildcardParameters(
+                        GenericTypeReflector.erase(actualBeanType)));
+                }
+                return exactReturnType;
+            }
+
             @Override
             public <A extends Annotation> Optional<A> getAnnotation(Class<A> anno) {
                 return annoCache.computeIfAbsent(anno, x ->
```

**Comment by stevenschlansker** [comment]
Ran into another issue while trying to figure out a fix here: https://github.com/leangen/geantyref/issues/13

**Comment by stevenschlansker** [comment]
Hi @martin-sladecek , this should be fixed in 3.24.1. Thanks again for reporting this :smile_cat:
